core/src/util_regex.cpp

```diff
@@ -0,0 +1,343 @@
+  const inputstr = Module.UTF8ToString(input);
+  const re = new RegExp(patternstr);
+  const result = re.exec(inputstr);
+  if (!result) return 0; // no match
+  const g1 = result[1];
+  if (DEBUG_JS) console.dir({patternstr,inputstr,re,result,g1});
+  if (!g1) return 0;
+  return stringToNewUTF8(g1);
+});
+
+
+/**
+ * RegexSubstitute
+ * @param pattern the string, sans trailing $, for the pattern
+ * @param input the text to match against
+ * @param to the replacement text
+ * @return the entire updated output string
+ */
+EM_JS(char*, RegexSubstitute, (const char* pattern, const char *input, const char *to), {
+  const DEBUG_JS = false;
+  if (!pattern) return -1;
+  if (!input) return -1;
+  const patternstr = Module.UTF8ToString(pattern) + '$';
+  const inputstr = Module.UTF8ToString(input);
+  const tostr = Module.UTF8ToString(to);
+  const re = new RegExp(patternstr);
+  const output = inputstr.replace(re, tostr);
+  return stringToNewUTF8(output);
+});
+
+// pull in the generated table
+#include "util_normalize_table.h"
+
+#endif
+
+
+namespace km {
+namespace core {
+namespace util {
+
+/** find the */
+int32_t km_regex::findIndex(const std::u32string &match, const std::deque<std::u32string> &list) {
+  int32_t index = 0;
+  for(auto e = list.begin(); e < list.end(); e++, index++) {
+    if (match == *e) {
+      return index;
+    }
+  }
+  return -1; // not found
+}
+
+km_regex::km_regex()
+#if KMN_NO_ICU
+#else
+ : fPattern(nullptr)
+#endif
+{
+
+}
+
+
+km_regex::km_regex(const km_regex& other)
+#if KMN_NO_ICU
+ : fPattern(other.fPattern)
+#else
+ : fPattern(nullptr)
+#endif
+{
+#if KMN_NO_ICU
+
+#else
+    if (other.fPattern) {
+    // clone pattern
+    fPattern.reset(other.fPattern->clone());
+  }
+#endif
+}
+
+km_regex::km_regex(const std::u32string &pattern)
+#if KMN_NO_ICU
+  : fPattern(pattern)
+#else
+  : fPattern(nullptr)
+#endif
+{
+  init(pattern);
+}
+
+km_regex::~km_regex() {
+
+}
+
+bool km_regex::valid() const {
+#if KMN_NO_ICU
+  return (!fPattern.empty());
+#else
+  // valid if fPattern is present.
+  return !!fPattern;
+#endif
+}
+
+bool km_regex::init(const std::u32string &pattern) {
+#if KMN_NO_ICU
+// for now- new regex every time.
+  assert(!pattern.empty());
+  fPattern = pattern;
+  return true; // TODO
+#else
+  if (pattern.empty()) {
+    return false;
+  }
+  // TODO-LDML: if we have mapFrom, may need to do other processing.
+  std::u16string patstr = km::core::kmx::u32string_to_u16string(pattern);
+  UErrorCode status           = U_ZERO_ERROR;
+  /* const */ icu::UnicodeString patustr = icu::UnicodeString(patstr.data(), (int32_t)patstr.length());
+  // add '$' to match to end
+  patustr.append(u'$'); // TODO-LDML: may need to escape some markers. Marker #91 will look like a `[` to the pattern
+  fPattern.reset(icu::RegexPattern::compile(patustr, 0, status));
+  return (UASSERT_SUCCESS(status));
+#endif
+}
+
+size_t km_regex::apply(const std::u32string &input, std::u32string &output,
```

Comment:
Well, if you're targetting Linux, Mac, or Windows you don't have a JS runtime.. so yes the ICU is needed!

Also, the ICU form caches the regex so will be more performant at the moment.

It could be noted (at least if it comes up a trivia party game) to note that part of the JS code _is actually using ICU inside_ when running in a browser or Node.js context.  So the equivalent of the ICU code gets invoked. (At least for the normalization- v8 has an optimized regex engine.)

Reply:
The ICU side has more comments :D

Reply:
Where I was going with my "intuition" bit - I'd expect the non-JS side to tie in neatly with ICU stuff, while the WASM/JS part would need wrappers and extra handling to force the data into the proper form in each direction.

Even if the JS side does have browser-included ICU, that's at least a layer away.  Unlike sd the intermediate layer(s) happen to do exactly what we need to do with ICU code outside of JS-land, I'd expect the JS version to be larger.  (Its section was developed later and thus is not the internal standard.)

Hence my statement of surprise and continued concern here.

Reply:
well, the languages are just different.  Regex is a first class language syntax feature in JS.  I can write `[all,first,second] = /([a-z]+)([0-9]+)/.exec('abcd1234')
` in JS where in C++ I have to

- allocate the input string
- check for memory errors
- construct the regex matcher object
- check for memory or syntax errs
- execute the match
- yes, check errs again
- check the number of resulting objects
- copy the three strings somewhere

so, it's not that surprising to me.